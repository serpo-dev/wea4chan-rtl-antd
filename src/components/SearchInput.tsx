import { AutoComplete, Button, Form, message } from "antd";
import { useState, useEffect, FC } from "react";
import { getCity } from "../async/getCity";
import { useNavigate } from "react-router-dom";

interface Option {
    label: string;
    value: string;
}

export const SearchInput: FC = () => {
    const [options, setOptions] = useState<Option[] | undefined>(undefined);
    const [error, setError] = useState<string | null>(null);

    const [form] = Form.useForm();
    const [messageApi, contextHolder] = message.useMessage();
    const navigate = useNavigate();

    async function handleSearch(value: string) {
        const res = await getCity(value);
        if (res && res.length > 0) {
            const opts = res?.map((e) => {
                return {
                    label: e.name,
                    value: e.latitude + "; " + e.longitude,
                };
            });
            setOptions(opts);
        }
    }

    function onFinish({ value }: { value: string }) {
        if (value) {
            const trimmedValue = value.trim();

            const regex = /-{0,}\d+\.\d+;.{0,}-{0,}\d+/;
            const match = trimmedValue.match(regex);

            if (match?.[0] !== trimmedValue) {
                setError(`You need to enter the coordinates in format "x; y"`);
                return;
            }

            const coordinates = trimmedValue.split(";");
            const city = options?.filter((e) => e.value === trimmedValue)[0][
                "label"
            ];
            navigate(
                `/q?latitude=${coordinates[0].trim()}&longitude=${coordinates[1].trim()}${
                    city ? "&city=" + city?.trim() : ""
                }`
            );
            return;
        }
        return;
    }

    useEffect(() => {
        if (error !== null) {
            messageApi.open({ type: "error", content: error });
            setError(null);
        }
    }, [error, messageApi]);

    return (
        <div style={{ display: "flex", flexDirection: "column" }}>
            {contextHolder}
            <Form
                form={form}
                layout="horizontal"
                onFinish={onFinish}
                style={{ display: "flex", flexDirection: "row" }}
            >
                <Form.Item name="value">
                    <AutoComplete
                        size="large"
                        style={{
                            width: 400,
                        }}
                        onSearch={handleSearch}
                        placeholder="Search for coordinates..."
                        options={options}
                    />
                </Form.Item>
                <Form.Item>
                    <Button
                        title="Search button"
                        size="large"
                        type="primary"
                        htmlType="submit"
                    >
                        Search
                    </Button>
                </Form.Item>
            </Form>
        </div>
    );
};
