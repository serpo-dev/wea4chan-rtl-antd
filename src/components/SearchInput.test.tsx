import {
    cleanup,
    fireEvent,
    render,
    screen,
    waitFor,
} from "@testing-library/react";
import "@testing-library/jest-dom";
import { SearchInput } from "./SearchInput";
import { BrowserRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";

const mockedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
    ...(jest.requireActual("react-router-dom") as any),
    useNavigate: () => mockedNavigate,
}));

describe("should render SearchInput properly", () => {
    afterEach(cleanup);
    afterEach(() => mockedNavigate.mockRestore());

    it("should match the snapshot", () => {
        const { asFragment } = render(<SearchInput />, {
            wrapper: BrowserRouter,
        });
        expect(asFragment()).toMatchSnapshot();
    });

    it("should be redirect button", async () => {
        render(<SearchInput />, { wrapper: BrowserRouter });
        const searchButton = screen.getByTitle("Search button");
        expect(searchButton).toBeInTheDocument();
    });

    it("should be input field", async () => {
        render(<SearchInput />, { wrapper: BrowserRouter });
        let input = screen.getByRole<HTMLInputElement>("combobox");
        expect(input).toBeInTheDocument();
    });

    it("should find the coordinates by typing a name of the city", async () => {
        render(<SearchInput />, { wrapper: BrowserRouter });
        let input = screen.getByRole<HTMLInputElement>("combobox");
        userEvent.type(input, "yekaterinb");
        const option = await screen.findByText("Yekaterinburg");
        userEvent.click(option);
        await waitFor(() => expect(input.value).toEqual("56.8519; 60.6122"));
    });

    it("should redirect to the city page when the user presses 'enter'", async () => {
        render(<SearchInput />, { wrapper: BrowserRouter });

        const input = screen.getByRole<HTMLInputElement>("combobox");
        userEvent.type(input, "56.8519; 60.6122{enter}");

        const resultUrl = "/q?latitude=56.8519&longitude=60.6122";
        await waitFor(() => {
            expect(mockedNavigate).toHaveBeenCalledWith(resultUrl);
        });
    });

    it("should redirect to the city page when the user presses on 'Search' button", async () => {
        render(<SearchInput />, { wrapper: BrowserRouter });

        const input = screen.getByRole<HTMLInputElement>("combobox");
        userEvent.type(input, "56.8519; 60.6122");
        const searchButton = screen.getByTitle("Search button");
        fireEvent.click(searchButton);

        const resultUrl = "/q?latitude=56.8519&longitude=60.6122";
        await waitFor(() => {
            expect(mockedNavigate).toHaveBeenCalledWith(resultUrl);
        });
    });
});
