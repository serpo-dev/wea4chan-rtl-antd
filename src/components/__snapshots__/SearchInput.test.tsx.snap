// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SearchInput properly should match the snapshot 1`] = `
<DocumentFragment>
  <div
    style="display: flex; flex-direction: column;"
  >
    <form
      class="ant-form ant-form-horizontal css-dev-only-do-not-override-12jzuas"
      style="display: flex; flex-direction: row;"
    >
      <div
        class="ant-form-item css-dev-only-do-not-override-12jzuas"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-12jzuas"
        >
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-12jzuas"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-select ant-select-lg ant-select-in-form-item ant-select-auto-complete css-dev-only-do-not-override-12jzuas ant-select-single ant-select-show-search"
                  style="width: 400px;"
                >
                  <div
                    class="ant-select-selector"
                  >
                    <span
                      class="ant-select-selection-search"
                    >
                      <input
                        aria-activedescendant="value_list_0"
                        aria-autocomplete="list"
                        aria-controls="value_list"
                        aria-expanded="false"
                        aria-haspopup="listbox"
                        aria-owns="value_list"
                        autocomplete="off"
                        class="ant-select-selection-search-input"
                        id="value"
                        role="combobox"
                        type="search"
                        value=""
                      />
                    </span>
                    <span
                      class="ant-select-selection-placeholder"
                    >
                      Search for coordinates...
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-form-item css-dev-only-do-not-override-12jzuas"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-12jzuas"
        >
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-12jzuas"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <button
                  class="ant-btn css-dev-only-do-not-override-12jzuas ant-btn-primary ant-btn-lg"
                  title="Search button"
                  type="submit"
                >
                  <span>
                    Search
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
